File: 0112-path-sum/0112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False

        ans = self.traverse(root, False, 0, targetSum)
        return ans

    def traverse(self, curr, ans, sum, targetSum):
        newSum = curr.val + sum

        if curr.left == None and curr.right == None:
            if newSum == targetSum:
                ans = True
                return ans
        
        leftAns = False
        rightAns = False
        if curr.left != None:
           leftAns = self.traverse(curr.left, ans, newSum, targetSum)
        
        if curr.right != None:
            rightAns = self.traverse(curr.right, ans, newSum, targetSum)
        
        return leftAns or rightAns

File: 0112-path-sum/0112-path-sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */
var hasPathSum = function (root, targetSum) {
    let ans = false;

    const traverse = function(curr, sum) {
        let newSum = curr.val + sum;

        if(!curr.left && !curr.right) {
            if(newSum === targetSum) {
                ans = ans || true;
                return;
            }
        }

        curr.left && traverse(curr.left, newSum);
        curr.right && traverse(curr.right, newSum);
    }

    root && traverse(root, 0);

    return ans;
};